### DISCUSSION TITLE
Execution from android

### DISCUSSION THREAD

**Discussion by rdluffy** [termux/termux-app#3372] [Q&A] [answered]
Hello, 
I'm wondering how Termux is able to execute a binary that is in /data/local/tmp/xxx/app but when I'm trying to do it in a standard android application via (Runtime.getRuntime().exec(...)) it tells me "permission denied".
I have the same issue when the binary is in my directory (/data/data/packagename/files/app) and it is working for termux.

If someone can give me an insight ?

Thanks.

**Comment by agnostic-apollo** [comment, marked as answer]
Sorry, that's a top secret.  We don't tell others.

But on a serious note, your app needs to use `targetSdkVersion` `<= 28` to allow execution from app data directory, unless executable file is shipped inside apk file as native library, check `android-10` branch or https://github.com/agnostic-apollo/TaskerAppFactory as sample. The `/data/local/tmp` directory is owned by `adb` `shell` user and you will only be allowed to execute if file has `777` permissions, app users normally can't read/write files in that directory. Make sure to restart device after lowering `targetSdkVersion`.

https://github.com/termux/termux-app/issues/2445#issuecomment-985484952

https://github.com/termux/termux-app/issues/2155
